### PR TITLE
Updating winry315 to use ENCODER_MAP_ENABLE

### PR DESCRIPTION
**Pull request by andrewinmoco**
## Description

Updating winry315 to use ENCODER_MAP_ENABLE. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/winry/winry315/keymaps/via/config.h [hunk 1]
```diff
@@ -1,6 +0,0 @@
-// Copyright 2022 Sergey Vlasov (@sigprof)
-// SPDX-License-Identifier: GPL-2.0-or-later
-
-#pragma once
-
-#define DYNAMIC_KEYMAP_LAYER_COUNT 8
```
Comment (waffle87) [inline keyboards/winry/winry315/keymaps/via/config.h:6]: Why is this being removed?

File: keyboards/winry/winry315/keymaps/via/keymap.c [hunk 2]
```diff
@@ -3,42 +3,37 @@
...
-             KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
-             KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,
-             KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS
+
+    // RGB configuration
+    [_RGB] = LAYOUT_top(
+            RGB_M_P,      RGB_M_B,     RGB_M_R,
+        RGB_MOD, RGB_HUI, RGB_SAI, RGB_VAI, RGB_SPI,
+        RGB_RMOD,RGB_HUD, RGB_SAD, RGB_VAD, RGB_SPD,
+        KC_TRNS, RGB_TOG, RGB_M_P, RGB_M_B, RGB_M_R
```
Comment (waffle87) [inline keyboards/winry/winry315/keymaps/via/keymap.c:30]: Above comment w.r.t removing layers.

File: keyboards/winry/winry315/rules.mk [hunk 3]
```diff
@@ -11,5 +11,6 @@ BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
 RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
 AUDIO_ENABLE = no           # Audio output
 ENCODER_ENABLE = yes
+ENCODER_MAP_ENABLE = yes
```
Comment (waffle87) [inline keyboards/winry/winry315/rules.mk:14]: This should not be enabled at the keyboard level, and should be moved to `keymaps/*/rules.mk`.

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/winry/winry315/keymaps/via/keymap.c (hunk 2 above)
Comment (sigprof) [reply keyboards/winry/winry315/keymaps/via/keymap.c:30]: Actually this part may be OK — the keymap introspection code will provide the required number of layers completely filled with `KC_TRNS` even if they are not explicitly initialized in `keymaps`, so the empty layers at the end of the `via` keymap may be safely omitted.

**Review by sigprof — CHANGES_REQUESTED** [review]
The main concern with these changes is that they also require a corresponding update of the [VIA JSON for this board](https://github.com/the-via/keyboards/blob/master/v3/winry/winry315.json), and that would break compatibility with the older firmware versions, forcing all users to reflash the firmware to restore the VIA functionality.  I'm not 100% sure what would be the proper way here — maybe the USB IDs should be changed, so that both the old and the new VIA JSON versions may be kept; maybe VIA should get proper support for `firmwareVersion` first (with the possibility to keep multiple versions of JSONs for the same board in the VIA repo); or maybe we should just ignore the existence of older firmware versions (this includes the factory firmware — apparently the AliExpress sellers now advertise this board as “working with VIA out of the box”, because the upstream VIA JSON is compatible with the factory firmware too).

In addition, if we are breaking the compatibility by removing the old encoder actions support, the extra `null` entries in `direct_pins` may be removed too — these entries were added only to make the encode actions work, and are useless when the core encoder mapping support is enabled.

File: keyboards/winry/winry315/rules.mk (hunk 3 above)
Comment (andrewinmoco) [reply keyboards/winry/winry315/rules.mk]: I've made this change.

**Comment by andrewinmoco** [comment]
> The main concern with these changes is that they also require a corresponding update of the [VIA JSON for this board](https://github.com/the-via/keyboards/blob/master/v3/winry/winry315.json), and that would break compatibility with the older firmware versions, forcing all users to reflash the firmware to restore the VIA functionality. I'm not 100% sure what would be the proper way here — maybe the USB IDs should be changed, so that both the old and the new VIA JSON versions may be kept; maybe VIA should get proper support for `firmwareVersion` first (with the possibility to keep multiple versions of JSONs for the same board in the VIA repo); or maybe we should just ignore the existence of older firmware versions (this includes the factory firmware — apparently the AliExpress sellers now advertise this board as “working with VIA out of the box”, because the upstream VIA JSON is compatible with the factory firmware too).
 

Working with the VIA discord I was able to make an [updated VIA JSON](https://github.com/andrewinmoco/keyboards/blob/master/src/winry/winry315.json).   When I tested the current JSON with the current VIA firmware, everything but the encoders.   If we did break backwards compatibility, only the encoders would fail.  

I'm happy to make changes/work towards whatever path makes sense for the project, I'm just not informed enough to know which path that is. 


> In addition, if we are breaking the compatibility by removing the old encoder actions support, the extra `null` entries in `direct_pins` may be removed too — these entries were added only to make the encoder actions work, and are useless when the core encoder mapping support is enabled.

I've updated the config files to remove these references.

File: keyboards/winry/winry315/keymaps/via/config.h (hunk 1 above)
Comment (andrewinmoco) [reply keyboards/winry/winry315/keymaps/via/config.h]: I've restored the layers.

File: keyboards/winry/winry315/keymaps/via/keymap.c (hunk 2 above)
Comment (andrewinmoco) [reply keyboards/winry/winry315/keymaps/via/keymap.c:30]: I've tested this and they layers work in VIA even if they aren't created in the keymap file.

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Also, IIRC, there is some "via firmware version" that can be set to ensure that it only uses the correct version of the via json.

**Comment by andrewinmoco** [comment]
I've added a VIA_FIRMWARE_VERSION 1 to the JSON, and a VIA_FIRMWARE_VERSION to config.h 

In discussions on the VIA discord it was said that the firmware version is only used for lighting right now (and doesn't current support encoders).  Adding it in in case this changes in the future.      The VIA team also said that if I push the new JSON to the v3 folder it'll be used by the new firmware since the new firmware uses VIA v3 and the older v2 firmwares will still use the older JSON in the SRC folder.

**Comment by zvecr** [comment]
> The VIA team also said that if I push the new JSON to the v3 folder it'll be used by the new firmware since the new firmware uses VIA v3 and the older v2 firmwares will still use the older JSON in the SRC folder.

Except they fail to capture the case of existing v3 firmwares that have been created before this change. Those will not work correctly at all, nor does VIA really have a way to solve the problem.

**Comment by andrewinmoco** [comment]
So, I'm unsure as to what (if anything) I can do to move this forward.    Are we at a point where (basically) until VIA comes up with versioning that supports the entire layout, that older firmware's can't be updated to use new/breaking features?     

If that's true then I'll likely just keep this change in my fork only or make my own firmware development easier.

**Comment by sigprof** [comment]
The VIA issue about the migration to encoder mapping is https://github.com/the-via/app/issues/155.  The QMK decision is to keep things as they are until the next breaking change (2023-05-28), and then replace matrix hacks with the proper encoder mapping *on the `develop` branch*, so that the changes would come to the `master` branch after yet another breaking change period (3 months).  Hopefully by that time the VIA team will do something about this problem (or, if that won't happen, the compatibility with older firmwares would just be broken when new VIA JSONs with the encoder mapping support would be merged into `the-via/keyboards`, so that any existing users would need to update the firmware in order to use VIA).

File: keyboards/winry/winry315/winry315.c
Comment (sigprof) [inline keyboards/winry/winry315/winry315.c:1]: The `#if !defined(VIA_ENABLE) && defined(ENCODER_ENABLE)` condition in this file can now be simplified to `#ifdef ENCODER_ENABLE`.

**Review by sigprof — COMMENTED** [review]
I completely agree with the direction of these changes, but they can't be merged just now because of VIA compatibility concerns.

**Review by drashna — APPROVED** [review]

**Pull request closed without merging**